### PR TITLE
Divide Util.hpp and fix some bugs there

### LINKED ISSUE #227: Util::matrixMatrixProduct produces a wrong result with BLAS
#### Before posting an issue

Search existing GitHub issues to make sure the issue does not already exist:
https://github.com/xanaduai/pennylane-lightning/issues

If posting a PennyLane-Lightning issue, delete everything above the dashed line, and fill
in the template. If the issue is a bug, start the title of the issue with [BUG].

If making a feature request, delete the following template and describe, in detail,
the feature and why it is needed.

For general technical details check out our documentation:
https://pennylane-lightning.readthedocs.io/

-------------------------------------------------------------------------------------------------------------

#### Issue description

Description of the issue - include code snippets and screenshots here
if relevant. You may use the following template below

* *Expected behavior:* (What you expect to happen)
C++ tests should pass with BLAS

* *Actual behavior:* (What actually happens)
Tests fail

* *Reproduces how often:* (What percentage of the time does it reproduce?)

* *System information:* (post the output of `import pennylane as qml; qml.about()`)

This is in fact caused by two convoluted problems:
1. `matrixMatProd(m_left, m_right, .., transpose=true)` function in `Util.hpp` transpose `m_right` (expected behaviour) when BLAS is not used, but BLAS version transpose `m_left`. 
2. Currently C++ tests do not use BLAS even when `-DENABLE_BLAS=ON` is given as `tests/CMakeLists.txt` does not include external library or compile flags set by the options.

Fortunately, as wheels are built with `DENABLE_BLAS=OFF`, this does not cause an issue for usual users (unless someone compiles from the source directly). Still, I will fix this ASAP.

### PR DESCRIPTION
**Pull request by chaeyeunpark**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** After several recent commits, `Util.hpp` file becomes >1000 lines, and some hidden bugs are found. This PR (1) divides this file into smaller files and (2) fix bugs.

**Description of the Change:** Divide `Util.hpp` into `LinearAlgebra.hpp`, `ConstantUtil.hpp`, and `BitUtil.hpp`. Also fixes bugs in `second_elts_of` and `matrixMatProd`. Also changed `CMakeLists.txt` and `Makefile`.

**Benefits:** Slightly better compile time.

**Possible Drawbacks:**

**Related GitHub Issues:**
Closes #227

### CODE REVIEW

File: pennylane_lightning/src/util/LinearAlgebra.hpp [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+        constexpr std::complex<T> co{1, 0};
+        constexpr std::complex<T> cz{0, 0};
+        const auto tr = (transpose) ? CblasTrans : CblasNoTrans;
+        if constexpr (std::is_same_v<T, float>) {
+            cblas_cgemm(CblasRowMajor, CblasNoTrans, tr, m, n, k, &co, m_left,
+                        k, m_right, transpose ? n : k, &cz, m_out, n);
+        } else if constexpr (std::is_same_v<T, double>) {
+            cblas_zgemm(CblasRowMajor, CblasNoTrans, tr, m, n, k, &co, m_left,
+                        k, m_right, transpose ? n : k, &cz, m_out, n);
+        }
```
Comment (chaeyeunpark) [inline pennylane_lightning/src/util/LinearAlgebra.hpp:640]: FYI, the previous code was 
https://github.com/PennyLaneAI/pennylane-lightning/blob/8b9d4127831a1bbe18b19fe5029c3bd2707b3a0b/pennylane_lightning/src/util/Util.hpp#L973-L979

File: pennylane_lightning/src/util/ConstantUtil.hpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+second_elts_of(const std::array<std::pair<T, U>, size> &arr) {
+    // TODO: change to std::transform in C++20
+    std::array<U, size> res = {
+        U{},
+    };
+    for (size_t i = 0; i < size; i++) {
+        res[i] = std::get<1>(arr[i]);
+    }
+    return res;
+}
```
Comment (chaeyeunpark) [inline pennylane_lightning/src/util/ConstantUtil.hpp:103]: FYI, this function was the same as `fist_elts_of` previously. Besides these two functions, every other functions are the same.

**Review by mlxd — APPROVED** [review]
Great job on this @chaeyeunpark 
That refactor was definitely needed (and nice work on the bug fixes too!)

**Review by maliasadi — APPROVED** [review]
Nice work! Also, thanks for fixing those bugs :slightly_smiling_face:

**Pull request merged**